src/translations/indexMapper.js

```diff
@@ -0,0 +1,362 @@
+import { arrayFilter, arrayMap } from './../helpers/array';
+import { getListWithRemovedItems, getListWithInsertedItems } from './maps/utils/visuallyIndexed';
+import { rangeEach } from '../helpers/number';
+import IndexMap from './maps/indexMap';
+import SkipMap from './maps/skipMap';
+import MapCollection from './mapCollection';
+import localHooks from '../mixins/localHooks';
+import { mixin } from '../helpers/object';
+import { isDefined } from '../helpers/mixed';
+
+class IndexMapper {
+  constructor() {
+    /**
+     * Map storing the sequence of indexes.
+     *
+     * @type {IndexMap}
+     */
+    this.indexesSequence = new IndexMap();
+    /**
+     * Collection for different skip maps. Indexes marked as skipped in any map won't be rendered.
+     *
+     * @type {MapCollection}
+     */
+    this.skipCollection = new MapCollection();
+    /**
+     * Collection for another kind of mappings.
+     *
+     * @type {MapCollection}
+     */
+    this.variousMappingsCollection = new MapCollection();
+    /**
+     * Cache for skip result for particular indexes.
+     *
+     * @type {Array}
+     */
+    this.flattenSkipList = [];
+    /**
+     * Cache for list of not skipped indexes, respecting the indexes sequence.
+     *
+     * @type {Array}
+     */
+    this.notSkippedIndexesCache = [];
+    /**
+     * Flag determining whether operations performed on index mapper were batched.
+     *
+     * @type {Boolean}
+     */
+    this.isBatched = false;
+    /**
+     * Flag determining whether any action on indexes sequence or skipped indexes was performed.
+     *
+     * @type {Boolean}
+     */
+    this.cachedIndexesChange = false;
+
+    this.indexesSequence.addLocalHook('change', () => {
+      this.cachedIndexesChange = true;
+
+      // Sequence of visible indexes might change.
+      this.updateCache();
+
+      this.runLocalHooks('change', this.indexesSequence, null);
+    });
+
+    this.skipCollection.addLocalHook('change', (changedMap) => {
+      this.cachedIndexesChange = true;
+
+      // Number of visible indexes might change.
+      this.updateCache();
+
+      this.runLocalHooks('change', changedMap, this.skipCollection);
+    });
+
+    this.variousMappingsCollection.addLocalHook('change', (changedMap) => {
+      this.runLocalHooks('change', changedMap, this.variousMappingsCollection);
+    });
+  }
+
+  /**
+   * Execute batch operations with updating cache.
+   *
+   * @param {Function} curriedBatchOperations Batched operations curried in a function.
+   */
+  executeBatchOperations(curriedBatchOperations) {
+    const actualFlag = this.isBatched;
+
+    this.isBatched = true;
+
+    curriedBatchOperations(this);
```

Comment:
The idea was that you can use arrow function and have both, current plugin context and indexMapper:

```ts
this.rowIndexMapper.executeBatchOperations((indexMapper) => {
  // this is this plugin context
  // indexMapper is the indexMapper, wow!
}
```

But it's the same as:

```ts
this.rowIndexMapper.executeBatchOperations(() => {
  // this is the plugin context
  // so... index mapper is this.rowIndexMapper, tada!
}
```

And since currently, the preferred use case has changed to favor the use of public API `this.hot.executeBatchOperations` which calls `executeBatchOperations` for both `rowIndexMapper` and `columnIndexMapper` this can be confusing and misleading.

Maybe we should discuss the use case scenarios and we might be missing a test here to check if the callback was called with the correct context.

https://github.com/handsontable/handsontable/blob/ef1be8085234d10b9c2341dd8fc96806f50e72e2/src/core.js#L878-L884

What will be the `this` in `curriedBatchOperations` callback? `hotInstance` ? 🕵

Reply:
I wonder, what is the usefulness of the reference to the `indexMapper` inside of the callback? Knowing this will help me better understand the whole thing. 

I am genuinely curious, as plugin hook callbacks did not have a reference to the `pluginHook` class.

Reply:
Let's say we have to call multiple operations on the exact same mapper:
```ts
this.rowIndexMapper.executeBatchOperations(() => {
    this.rowIndexMapper.removeIndexes();
    this.rowIndexMapper.insertIndexes();
    this.rowIndexMapper.moveIndexes();
});
```

With `this` passed as an argument it could be used without arrow function:
```ts
this.rowIndexMapper.executeBatchOperations(function (mapper) {
    mapper.removeIndexes();
    mapper.insertIndexes();
    mapper.moveIndexes();
});
```

It wasn't introduced for the potential future. That was the original idea. But the main use case has changed since then and I don't think it's that much relevant anymore.

Reply:
It seems to be useless 🤔.

> What will be the `this` in `curriedBatchOperations` callback? `hotInstance` ?

Context in which it was used. If we use it inside the plugin `this` reference to a plugin instance. It have reference to the `hot` and `hot` have properties for index mappers. I will remove it. Thanks @warpech.